### PR TITLE
Drop AuthenticatedClientSessionStore from user sessions

### LINKED ISSUE #30934: Drop `AuthenticatedClientSessionStore` from user sessions
### Description

`AuthenticatedClientSessionStore` is a field of user sessions that maps the user session to its client sessions.
The client session can store the user session it belongs to and we can search the embedded and/or remote Infinispan caches to list all the sessions belonging to a single user session.

### Discussion

https://github.com/keycloak/keycloak/discussions/30532

### Motivation

`AuthenticatedClientSessionStore` is a source of conflict if 2 or more client sessions need to be linked to a single user session. 

### Details

_No response_

### PR DESCRIPTION
**Pull request by pruivo**
New entities for client and user sessions, more query friendly. The client sessions are found using query instead of storing them in the user session entity.
Remove of sessions by its field is done based on queries.

Closes #30934

### CODE REVIEW

**Review by ryanemerson — COMMENTED** [review]
A few general comments, I can't really comment the on the semantics of dropping `AuthenticatedClientSessionStore` as I don't have enough knowledge of the area.

I noticed in my IDE that `AuthenticatedClientSessionStore` is still used, is this expected or an oversight?

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/QueryHelper.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/QueryHelper.java:127]: Yes, the publisher is only available in ISPN 15.1. I can add a comment.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/MultipleConditionQueryRemover.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/MultipleConditionQueryRemover.java:75]: Correct. Keycloak code is blocking and an instance is created per request.

**Comment by pruivo** [comment]
@ryanemerson `AuthenticatedClientSessionStore` is still used for single site deployment (embedded caches). I only created more "remote cache" (and query) friendly entities.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/QueryHelper.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/QueryHelper.java]: I changed the code to use `RemoteQuery.executeAsync()` and `RemoteQuery.executeStatementAsync()`. It is not a public API but we can access it. The `BatchingIterator` needs to be around.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/RemoteAuthenticatedClientSessionEntity.java
Comment (ryanemerson) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/RemoteAuthenticatedClientSessionEntity.java:89]: It saves a couple of lines of code if you use a constructor, that's the only advantage. I'm happy for you to keep as you prefer :+1:

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ahus1) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:203]: Looking at the configurations on KCB, I don't see a hint in the cache configurations that they should be indexed. Is that optional, but recommended? Looking at the comment "We need indexed caches because the delete statement fails for non-indexed cache." is seems that the comment might be outdated, or at least it confuses me as I don't see indexes on the KCB Infinispan scripts.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java
Comment (ahus1) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java:31]: When we use Ickle queries, how to these statements perform with an increasing number of user sessions and client sessions? On which queries do we use indexes, on which queries will we do a full scan of all entries in memory or in store (I know we don't use a persistent store in Infinispan yet)?

**Review by ahus1 — COMMENTED** [review]
Thank you for this PR. See below for some high-level questions I would like to understand before merging this. 

With the remote cache configuration getting more complicated in every iteration, I wonder if Keycloak should check the remote cache configuration at startup and report warnings or even fail to start it sees a misconfiguration. 

Keycloak already has great powers on Infinispan to upload the schema. With great powers comes great responsibility.  :spider_web:  

I know we discussed earlier that Keycloak configuring a production Infinispan shouldn't happen due to additional configuration options. But IMHO validating the config could safe us some trouble in development and also some tears for users trying to set it up in production.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:203]: The comment refers to https://issues.redhat.com/browse/ISPN-16270

We updated Infinspan to 15.0.7.Final this week and it makes the comment invalid. I'll revert these changes and create basic and non-indexed caches.

Note that the integration test suite uses this method and is not supposed to be used in production. It is useful for demos since we can start the Infinispan server with the default config :+1: 

For Keycloak 26, indexing is optional but I wouldn't recommend it. Only removing realms uses queries to remove the authentication sessions and login failures. I don't expect removing realms to be such a common operation or being performance sensitive. Also, I don't expect these 2 caches to have a large data set.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java:31]: I annotated all fields used in these queries to be indexed. If indexes are enabled in the external Infinispan, it will perform indexed queries. If not, it will do full scans. Hybrid queries should not happen at all. 

For the performance, we have to test it for KC 27 (volatile sessions won't be supported in KC 26, so this code will never be invoked). 
In the support case I worked on a couple of months ago, they were using indexed queries. With a total of 1'800'000 entries and a 6 nodes cluster (~300'000 entries per node), the indexed queries took an average of 7ms on each node. Non-indexed queries (full scans) were taking closer to 100ms if my memory is not failing me right now.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:203]: I've created https://github.com/pruivo/keycloak/pull/4 to clarify the docs about which version of Infinispan to use. Also consider mentioning this in the migration guide. 

I know this method is used in the integration tests, still a production environment shouldn't deviate much from it. And this is something we could automate, not necessarily in this issue though. 

Looking at the code, I see that we will retrieve the client sessions connected to user sessions in several occasions during login / logout, possibly also during token validation. So I'd say "Only removing realms uses queries to remove the authentication sessions and login failures." is only a partial answer to my question.

**Comment by pruivo** [comment]
@ahus1 I would love to know what kind of misconfiguration you have in mind. On another note, the Hot Rod client is unable to access the server configuration. We would need to REST client to fetch the cache configuration for validation. 

Anything that crosses my mind should work fine with keycloak: sync/async, indexed or not, with or without eviction, with or without persistence...

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java:31]: Thank you for these numbers. I read them that we would see a noticeable delay once we cross 100_000 entries per node. 

Given those numbers, I suggest to create a follow-up issue about indexing, and then merge the PR as is.

**Comment by ahus1** [comment]
> I would love to know what kind of misconfiguration you have in mind. 

Originally I had in mind that indexing must be enabled as I assume it would just not work without that, but that is now resolved after your previous comment. 

I see that in we describe in our docs to have `SYNC` mode, encoding of Protostream, and transaction of NON_XA pessimistic. `SYNC` mode is the only one we tested and support, and Protostream is probably essential. Not sure about the transaction type. 

The other thing that we also state in our docs is "if setting up `maxCount` for those caches used by persistent sessions, then also configure `owners` to be `1` for those". 

> On another note, the Hot Rod client is unable to access the server configuration. We would need to REST client to fetch the cache configuration for validation.

I assumed that we were using that interface when uploading the schema, but looking at the code we're using just the Remote Cache Manager for that. 

If you think such a check would make sense, let's create a follow-up issue for it. If you think it is not worth it, I'm ok to leave it as is, and wait for a feedback from users if they would need it.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:203]: You have answered the the question about the indexing of for the client sessions cache below and that part is now resolved. 

This then leaves only the documentation change PR plus the thing about the non-indexed basic caches open.

**Comment by pruivo** [comment]
In my opinion, document the correct configuration to use but do not restrict the user (by code, without a workaround)  to experiment with some different (aka crazy) configurations. 

ProtoStream encoding: it is only a requirement for caches that need a query. For simplicity, I would say it is a requirement for all caches. 

### Supported caches in Keycloak 26
* `work` -> Mostly events with unique keys, any configuration should work fine. 
* `authenticationSessions` -> unique keys, no conflicts, any configuration should work fine too.
* `actionTokens` -> same as `authenticationSessions` 
* `loginFailures` -> conflicts may happen, use a more strict configuration. Recommend `NON_XA` + `PESSIMISTIC`. Indexing is optional.

### Unsupported caches in Keycloak 26
All the user session and client session caches, same as `loginFailures`. Except for indexing, which we should recommend for large datasets. We need some testing to define what "large dataset" means, in other words, when indexing outperforms the full scan.

### Eviction in Keycloak 26
I wouldn't mention it since the supported caches are usually small. 

### Infinispan Persistence in Keycloak 26
Same as eviction. If my assumption is incorrect and these caches can grow quite a bit, we can add these topics to some advanced/sizing guides.

**Comment by ahus1** [comment]
I've tested this and I found that the listing of all client sessions occurs only during logout. All other steps during login and token refresh never trigger a search in the remote store.

**Review by ahus1 — APPROVED** [review]
Thank you, this is a big step forward. Thank you for this change.

**Pull request merged**
